Client/WarFare/UIItemUpgrade.cpp

```diff
@@ -0,0 +1,1289 @@
+﻿// UIItemUpgrade.cpp: implementation of the CUIItemUpgrade class.
+//Author : Monzantys(Mervan)
+//////////////////////////////////////////////////////////////////////
+
+#include "stdafx.h"
+#include "LocalInput.h"
+#include "APISocket.h"
+#include "GameProcMain.h"
+#include "UIItemUpgrade.h"
+#include "UIImageTooltipDlg.h"
+#include "UIInventory.h"
+#include "UIManager.h"
+#include "PlayerMySelf.h"
+#include "text_resources.h"
+#include "UIHotKeyDlg.h"
+#include "UISkillTreeDlg.h"
+#include "resource.h"
+#include <unordered_set>
+
+#ifdef _DEBUG
+#undef THIS_FILE
+static char THIS_FILE[] = __FILE__;
+#endif
+
+//////////////////////////////////////////////////////////////////////
+// Construction/Destruction
+//////////////////////////////////////////////////////////////////////
+
+CUIItemUpgrade::CUIItemUpgrade()
+{
+	for (int i = 0; i < MAX_ITEM_UPGRADE_SLOT; i++)
+	{
+		m_pMyUpgradeSLot[i] = nullptr;
+		m_iUpgradeSlotInvPos[i+1] = -1;
+	}
+
+	for (int i = 0; i < MAX_ITEM_INVENTORY; i++)
+	{
+		m_pMyUpgradeInv[i] = nullptr;
+		m_pBackupUpgradeInv[i] = nullptr;
+	}
+
+	m_pUpgradeResultSlot = nullptr;
+	m_pUpgradeItemSlot = nullptr;
+	m_iUpgradeSlotInvPos[0] = -1; //UpgradeItemSlot position
+	m_pUITooltipDlg = nullptr;
+	m_pStrMyGold = nullptr;
+}
+
+CUIItemUpgrade::~CUIItemUpgrade()
+{
+	Release();
```

Comment:
Avoid calling Release() in destructors. I understand this occurs elsewhere, but this is a problem, not something to be replicated.

Release() is a virtual method, which means that it typically doesn't - unintentionally - invoke the highest Release() implementation, only the CUIItemUpgrade::Release() implementation. While in this case it's technically fine, it's still a bad idea, so you should implement free-specific logic in the destructor explicitly.